### PR TITLE
Add some rust-based benchmarks

### PR DESCRIPTION
**Pull request by rkuris**
These are a little easier for me to wrap my head around and meet our immediate goals of knowing how fast inserts are both in large and small batches. They also make it a little easier for us to rerun with small changes.

### CODE REVIEW

File: grpc-testtool/benches/insert.rs [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+    // remove test directory
+    remove_dir_all(TESTDIR).ok();
+
+    // find the process server
+    let process_server = process_server_path().expect("Can't find process-server on path");
+    eprintln!("Using process-server {}", process_server.display());
+
+    // spawn a new one; use a separate thread to avoid zombies
+    std::thread::spawn(|| {
```
Comment (danlaine) [inline grpc-testtool/benches/insert.rs:61]: It seem we don't kill this at the end of the benchmarks. I feel like we should do that, no?

File: grpc-testtool/benches/insert.rs [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+            .arg("--config")
+            .arg(MERKLEDB_OPTIONAL_CONFIGURATIONS)
+            .process_group(0)
+            .spawn()
+            .expect("unable to start process-server")
+            .wait()
+    });
+
+    // wait for it to accept connections
+    retry("accept", || {
```
Comment (danlaine) [inline grpc-testtool/benches/insert.rs:76]:
```suggestion
    retry("can't connect to process server", || {
```

File: grpc-testtool/benches/insert.rs [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+                                    .sample_iter(&Alphanumeric)
+                                    .take(KEYLEN)
+                                    .collect::<Vec<u8>>(),
+                                rng.borrow_mut()
+                                    .sample_iter(&Alphanumeric)
+                                    .take(DATALEN)
+                                    .collect::<Vec<u8>>(),
+                            )
+                        })
+                        .map(|(key, value)| PutRequest { key, value })
```
Comment (danlaine) [inline grpc-testtool/benches/insert.rs:202]: Is this (calling map twice) considered more idiomatic than just doing one call to map like this?

```
                    let put_requests: Vec<PutRequest> = (0..s)
                        .map(|_| PutRequest {
                            key: rng
                                .borrow_mut()
                                .sample_iter(&Alphanumeric)
                                .take(KEYLEN)
                                .collect::<Vec<u8>>(),
                            value: rng
                                .borrow_mut()
                                .sample_iter(&Alphanumeric)
                                .take(DATALEN)
                                .collect::<Vec<u8>>(),
                        })
                        .collect();
```

File: grpc-testtool/benches/insert.rs [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    // We want a consistent seed, but we need different values for each batch, so we
+    // reseed each time we compute more data from the next seed value upwards
+    let seed = RefCell::new(0);
+
+    criterion.bench_with_input(
+        BenchmarkId::new("insert", BATCHSIZE),
+        &BATCHSIZE,
+        |b, &s| {
+            const KEYLEN: usize = 32;
+            const DATALEN: usize = 32;
```
Comment (danlaine) [inline grpc-testtool/benches/insert.rs:168]: Should these also be inputs to the benchmark? Could be done in a future PR

**Review by danlaine — CHANGES_REQUESTED** [review]

File: grpc-testtool/benches/insert.rs [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    Command::new("killall")
+        .arg("process-server")
+        .output()
+        .expect("killall should work");
+
+    // wait for them to die
+    retry("process server pids", || process_server_pids().is_empty());
+
+    // remove test directory
+    remove_dir_all(TESTDIR).ok();
```
[the remark this answers is not in the document]
Comment (rkuris) [reply grpc-testtool/benches/insert.rs:54]: This is a shortcut to toss the `Result`. I switched it to a more explicit syntax.

File: grpc-testtool/benches/insert.rs [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+/// Finds the first process-server on the path, or in some predefined locations
+///
+/// If the process-server isn't on the path, look for it in a target directory
+/// As a last resort, we check the parent in case you're running from the
+/// grpc-testtool directory
+fn process_server_path() -> Option<PathBuf> {
+    //
+    const OTHER_PLACES_TO_LOOK: &str = ":target/release:../target/release:.";
+
+    env::var_os("PATH").and_then(|mut paths| {
```
[the remark this answers is not in the document]
Comment (rkuris) [reply grpc-testtool/benches/insert.rs:129]: I'm going to assume everyone has a PATH, I mean, they ran this program without one? But yeah, if they don't, it won't look in those extra directories.

File: grpc-testtool/benches/insert.rs [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+                str::parse(line.trim_start().get(0..pos).unwrap_or_default()).unwrap_or_default()
+            })
+        })
+        .collect()
+}
+
+/// Finds the first process-server on the path, or in some predefined locations
+///
+/// If the process-server isn't on the path, look for it in a target directory
+/// As a last resort, we check the parent in case you're running from the
```
[the remark this answers is not in the document]
Comment (rkuris) [reply grpc-testtool/benches/insert.rs:128]: Fixed docs

File: grpc-testtool/benches/insert.rs [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+                        puts: put_requests,
+                        deletes: vec![],
+                    });
+
+                    // hand back the client and the request contents to the benchmark executor
+                    (client.clone(), req)
+                },
+                // this part is actually timed
+                |wbr| async {
+                    let (mut client, req) = wbr.await;
```
Comment (danlaine) [inline grpc-testtool/benches/insert.rs:219]: Does calling `await` here mean that we are including the time used to generate the inputs in the benchmark? When is the body of the async `setup` evaluated?

**Review by danlaine — APPROVED** [review]
Still not sure I understand [this bit](https://github.com/ava-labs/firewood/pull/528/files#r1492884323) but no need to block this

File: grpc-testtool/benches/insert.rs (hunk 8 above)
Comment (rkuris) [reply grpc-testtool/benches/insert.rs]: Yes! Good catch.

Fixed in latest version.

**Comment by rkuris** [comment]
> Still not sure I understand [this bit](https://github.com/ava-labs/firewood/pull/528/files#r1492884323) but no need to block this

The link doesn't really work so I'm not sure what part you're talking about, but I had added a dependency on jemallocator which we really dont need, so I removed it.

**Review by xinifinity — APPROVED** [review]

File: grpc-testtool/README.md [hunk 9]
```diff
@@ -21,29 +21,9 @@ There are 3 RPC specs that must be implemented:
 
 # Running
 
-To run a single test and make sure things are working, first check out and build the go and rust code.
-These have to be in the same directory. See the corresponding README for specific build requirements.
+To test the release version of firewood, just run `RUST_MIN_STACK=7000000 cargo bench`. If you make some changes and then
```
[the remark this answers is not in the document]
Comment (rkuris) [reply grpc-testtool/README.md:24]: No, but I experimented with different values and this one seems to let us run the longest. If you don't set this, the stack overflows.

[the remark this answers is not in the document]
Comment (rkuris) [reply grpc-testtool/README.md:24]: Created an issue to investigate. @xinifinity this is related to the IO subsystem and I'm going to guess this causes some performance losses to have the stack so large.

File: grpc-testtool/benches/insert.rs [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+/// The directory where the database will be created
+const TESTDIR: &str = "/tmp/benchdb";
+/// The port to use for testing
+const TESTPORT: u16 = 5000;
+/// The URI to connect to; this better match the TESTPORT
+const TESTURI: &str = "http://localhost:5000";
+/// Retry timeouts (in seconds); we want this long for processes
+/// to start and exit
+const RETRY_TIMEOUT_SEC: u32 = 5;
```
[the remark this answers is not in the document]
Comment (rkuris) [reply grpc-testtool/benches/insert.rs:26]: Created a ticket to address this as a followup

**Pull request merged**